### ISSUE TITLE
Running test.py: AttributeError: 'NoneType' object has no attribute 'get_num'

### ISSUE THREAD

**Issue by meier-johannes94** [chuangg/tdw-transport-challenge-starter-code#6]
We have done the required installation steps and the steps to run it locally. This means: Setting the `TRANSPORT_CHALLENGE` variable and then running 

```
# Run TDW with port 1071
./TDW/TDW.x86_64 -port 1071 &

# Run the python file
python test.py
```
We then get the following error message: 
```
Found path: /is/sg/jachterhold/tdw_build/TDW/TDW.x86_64
/is/sg/jachterhold/envs/tdw_py36/lib/python3.6/site-packages/sklearn/base.py:315: UserWarning: Trying to unpickle estimator GaussianMixture from version 0.22.2.post1 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
Controller connected
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    obs, info = env.reset(dataset[scene_number])
  File "/is/sg/jachterhold/threedworld/tdw-transport-challenge/tdw_transport_challenge/tdw_gym.py", line 116, in reset
    self.controller.init_scene(scene=scene, layout=layout, room = 5, random_seed=seed)
  File "/is/sg/jachterhold/threedworld/transport_challenge/transport_challenge/transport_controller.py", line 223, in init_scene
    return super().init_scene(scene=scene, layout=layout, room=room)
  File "/is/sg/jachterhold/envs/tdw_py36/lib/python3.6/site-packages/magnebot/magnebot_controller.py", line 432, in init_scene
    self._cache_static_data(resp=resp)
  File "/is/sg/jachterhold/threedworld/transport_challenge/transport_challenge/transport_controller.py", line 653, in _cache_static_data
    super()._cache_static_data(resp=resp)
  File "/is/sg/jachterhold/envs/tdw_py36/lib/python3.6/site-packages/magnebot/magnebot_controller.py", line 1867, in _cache_static_data
    for i in range(segmentation_colors.get_num()):
AttributeError: 'NoneType' object has no attribute 'get_num'
```
Here our pip freeze: 
```
altgraph==0.17.2
boto3==1.18.56
botocore==1.21.56
certifi==2021.5.30
charset-normalizer==2.0.6
cloudpickle==1.6.0
cycler==0.10.0
future==0.18.2
gym==0.18.0
idna==3.2
ikpy==3.2.2
imageio==2.9.0
importlib-metadata==4.8.1
jmespath==0.10.0
joblib==1.1.0
kiwisolver==1.3.1
magnebot==1.1.2
matplotlib==3.3.4
mpmath==1.2.1
numpy==1.19.5
opencv-python==4.5.3.56
packaging==21.0
pandas==1.1.5
Pillow==7.2.0
psutil==5.8.0
py-md-doc==0.2.4
-e git+https://github.com/hjweide/pyastar@f3a0559329cf50a140b166e03f3302b0f1f54310#egg=pyastar2d
pyglet==1.5.0
pyinstaller==4.5.1
pyinstaller-hooks-contrib==2021.3
pymongo==3.12.0
pyparsing==2.4.7
python-dateutil==2.8.2
pytz==2021.3
pyzmq==22.3.0
requests==2.26.0
s3transfer==0.5.0
scikit-learn==0.24.2
scipy==1.5.4
six==1.16.0
sympy==1.8
tdw==1.8.7.0
-e git+https://github.com/chuangg/tdw-transport-challenge@93ae780b2c7a54abaed66359cc277aa1e6ebace4#egg=tdw_transport_challenge
threadpoolctl==3.0.0
tqdm==4.62.3
-e git+https://github.com/alters-mit/transport_challenge.git@07ffe6ae87ee12c5488b42819b205b7eed39082f#egg=transport_challenge
typing-extensions==3.10.0.2
urllib3==1.26.7
zipp==3.6.0
zmq==0.0.0
```

How can we fix this?

**Comment by superhans21** [comment]
@abhi1092 
Thanks! Unfortunately we still have the same problem even after complete reinstallation: 
```
Found path: /is/sg/user/tdw_build/TDW/TDW.x86_64
/is/sg/user/envs/tdw_py36/lib/python3.6/site-packages/sklearn/base.py:315: UserWarning: Trying to unpickle estimator GaussianMixture from version 0.22.2.post1 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  UserWarning)
Controller connected
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    env.reset()
  File "/is/sg/user/envs/tdw_py36/lib/python3.6/site-packages/gym/wrappers/order_enforcing.py", line 16, in reset
    return self.env.reset(**kwargs)
  File "/is/sg/user/threedworld/tdw-transport-challenge/tdw_transport_challenge/tdw_gym.py", line 116, in reset
    self.controller.init_scene(scene=scene, layout=layout, room = 5, random_seed=seed)
  File "/is/sg/user/threedworld/transport_challenge/transport_challenge/transport_controller.py", line 223, in init_scene
    return super().init_scene(scene=scene, layout=layout, room=room)
  File "/is/sg/user/envs/tdw_py36/lib/python3.6/site-packages/magnebot/magnebot_controller.py", line 432, in init_scene
    self._cache_static_data(resp=resp)
  File "/is/sg/user/threedworld/transport_challenge/transport_challenge/transport_controller.py", line 653, in _cache_static_data
    super()._cache_static_data(resp=resp)
  File "/is/sg/user/envs/tdw_py36/lib/python3.6/site-packages/magnebot/magnebot_controller.py", line 1867, in _cache_static_data
    for i in range(segmentation_colors.get_num()):
AttributeError: 'NoneType' object has no attribute 'get_num'
```

```
altgraph==0.17.2
boto3==1.18.62
botocore==1.21.62
certifi==2021.10.8
charset-normalizer==2.0.7
cloudpickle==2.0.0
cycler==0.10.0
gym==0.21.0
idna==3.3
ikpy==3.1
imageio==2.9.0
importlib-metadata==4.8.1
jmespath==0.10.0
joblib==1.1.0
kiwisolver==1.3.1
magnebot==1.1.2
matplotlib==3.3.4
mpmath==1.2.1
numpy==1.19.5
opencv-python==4.5.3.56
pandas==1.1.5
Pillow==8.4.0
psutil==5.8.0
py-md-doc==0.2.4
pyastar==0.1
-e git+https://github.com/hjweide/pyastar@63ae69a8c65b168229f1c6ac8a58b753db5f5c18#egg=pyastar2d
pyinstaller==4.5.1
pyinstaller-hooks-contrib==2021.3
pymongo==3.12.0
pyparsing==2.4.7
python-dateutil==2.8.2
pytz==2021.3
pyzmq==22.3.0
requests==2.26.0
s3transfer==0.5.0
scikit-learn==0.24.2
scipy==1.5.4
six==1.16.0
sympy==1.9
tdw==1.8.7.0
-e git+https://github.com/chuangg/tdw-transport-challenge@93ae780b2c7a54abaed66359cc277aa1e6ebace4#egg=tdw_transport_challenge
threadpoolctl==3.0.0
tqdm==4.62.3
-e git+https://github.com/alters-mit/transport_challenge.git@ae3c8263a004ed651f15635ac17111858ad8bc4b#egg=transport_challenge
typing-extensions==3.10.0.2
urllib3==1.26.7
zipp==3.6.0
zmq==0.0.0
```

**Comment by alters-mit** [comment]
@meier-johannes94 

This is the problem:

```
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
```

There's something wrong with your X setup, which is causing the build to crash.

You might have xorg configured correctly. You might be launching the TDW build on the wrong display. Make sure that you follow the xorg setup instructions in this repo and in TDW's "Getting Started" doc.

**Comment by meier-johannes94** [comment]
@alters-mit 
We tried it on a second machine. My personal computer. And there we get the following Player.log. The messages also appeared on the other machine. 

```
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/libKernel32
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/libKernel32.so
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/Kernel32
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/libKernel32
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/libKernel32.so
Fallback handler could not load library /is/sg/user/tdw_build/TDW/TDW_Data/Mono/libKernel32
occurs.  
```

See here the full message: 
```
Mono path[0] = '/home/johannes/tdw_build/TDW/TDW_Data/Managed'
Mono config path = '/home/johannes/tdw_build/TDW/TDW_Data/MonoBleedingEdge/etc'
Preloaded 'libaudiopluginresonanceaudio.so'
Unable to preload the following plugins:
	libflexUtils.so
Display 0 'XWAYLAND0 15"': 1920x1080 (primary device).
Desktop is 1920 x 1080 @ 60 Hz
Initialize engine version: 2020.2.7f1 (c53830e277f1)
Plugins: Couldn't open OculusXRPlugin, error: OculusXRPlugin: cannot open shared object file: No such file or directory

[Subsystems] Discovering subsystems at path /home/johannes/tdw_build/TDW/TDW_Data/UnitySubsystems
[Subsystems] No descriptors matched for  examples in UnitySubsystems/OculusXRPlugin/UnitySubsystemsManifest.json.
[Subsystems] 1 'inputs' descriptors matched in UnitySubsystems/OculusXRPlugin/UnitySubsystemsManifest.json
[Subsystems] 1 'displays' descriptors matched in UnitySubsystems/OculusXRPlugin/UnitySubsystemsManifest.json
[Subsystems] No descriptors matched for  meshings in UnitySubsystems/OculusXRPlugin/UnitySubsystemsManifest.json.
[Subsystems] No descriptors matched for  examples in UnitySubsystems/WindowsMRXRSDK/UnitySubsystemsManifest.json.
[Subsystems] 1 'inputs' descriptors matched in UnitySubsystems/WindowsMRXRSDK/UnitySubsystemsManifest.json
[Subsystems] 1 'displays' descriptors matched in UnitySubsystems/WindowsMRXRSDK/UnitySubsystemsManifest.json
[Subsystems] 1 'meshings' descriptors matched in UnitySubsystems/WindowsMRXRSDK/UnitySubsystemsManifest.json
GfxDevice: creating device client; threaded=1
Renderer: Mesa Intel(R) UHD Graphics 630 (CFL GT2)
Vendor:   Intel
Version:  4.6 (Core Profile) Mesa 21.0.3
GLES:     0
 GL_3DFX_texture_compression_FXT1 GL_AMD_conservative_depth GL_AMD_depth_clamp_separate GL_AMD_draw_buffers_blend GL_AMD_gpu_shader_int64 GL_AMD_multi_draw_indirect GL_AMD_performance_monitor GL_AMD_pinned_memory GL_AMD_query_buffer_object GL_AMD_seamless_cubemap_per_texture GL_AMD_shader_stencil_export GL_AMD_shader_trinary_minmax GL_AMD_texture_texture4 GL_AMD_vertex_shader_layer GL_AMD_vertex_shader_viewport_index GL_ANGLE_texture_compression_dxt3 GL_ANGLE_texture_compression_dxt5 GL_ARB_ES2_compatibility GL_ARB_ES3_1_compatibility GL_ARB_ES3_2_compatibility GL_ARB_ES3_compatibility GL_ARB_arrays_of_arrays GL_ARB_base_instance GL_ARB_blend_func_extended GL_ARB_buffer_storage GL_ARB_clear_buffer_object GL_ARB_clear_texture GL_ARB_clip_control GL_ARB_compressed_texture_pixel_storage GL_ARB_compute_shader GL_ARB_compute_variable_group_size GL_ARB_conditional_render_inverted GL_ARB_conservative_depth GL_ARB_copy_buffer GL_ARB_copy_image GL_ARB_cull_distance GL_ARB_debug_output GL_ARB_depth_buffer_float GL_ARB
_depth_clamp GL_ARB_derivative_control GL_ARB_direct_state_access GL_ARB_draw_buffers GL_ARB_draw_buffers_blend GL_ARB_draw_elements_base_vertex GL_ARB_draw_indirect GL_ARB_draw_instanced GL_ARB_enhanced_layouts GL_ARB_explicit_attrib_location GL_ARB_explicit_uniform_location GL_ARB_fragment_coord_conventions GL_ARB_fragment_layer_viewport GL_ARB_fragment_shader GL_ARB_fragment_shader_interlock GL_ARB_framebuffer_no_attachments GL_ARB_framebuffer_object GL_ARB_framebuffer_sRGB GL_ARB_get_program_binary GL_ARB_get_texture_sub_image GL_ARB_gl_spirv GL_ARB_gpu_shader5 GL_ARB_gpu_shader_fp64 GL_ARB_gpu_shader_int64 GL_ARB_half_float_pixel GL_ARB_half_float_vertex GL_ARB_indirect_parameters GL_ARB_instanced_arrays GL_ARB_internalformat_query GL_ARB_internalformat_query2 GL_ARB_invalidate_subdata GL_ARB_map_buffer_alignment GL_ARB_map_buffer_range GL_ARB_multi_bind GL_ARB_multi_draw_indirect GL_ARB_occlusion_query2 GL_ARB_parallel_shader_compile GL_ARB_pipeline_statistics_query GL_ARB_pixel_buffer_object GL_ARB_po
int_sprite GL_ARB_polygon_offset_clamp GL_ARB_post_depth_coverage GL_ARB_program_interface_query GL_ARB_provoking_vertex GL_ARB_query_buffer_object GL_ARB_robust_buffer_access_behavior GL_ARB_robustness GL_ARB_sample_shading GL_ARB_sampler_objects GL_ARB_seamless_cube_map GL_ARB_seamless_cubemap_per_texture GL_ARB_separate_shader_objects GL_ARB_shader_atomic_counter_ops GL_ARB_shader_atomic_counters GL_ARB_shader_ballot GL_ARB_shader_bit_encoding GL_ARB_shader_clock GL_ARB_shader_draw_parameters GL_ARB_shader_group_vote GL_ARB_shader_image_load_store GL_ARB_shader_image_size GL_ARB_shader_objects GL_ARB_shader_precision GL_ARB_shader_stencil_export GL_ARB_shader_storage_buffer_object GL_ARB_shader_subroutine GL_ARB_shader_texture_image_samples GL_ARB_shader_texture_lod GL_ARB_shader_viewport_layer_array GL_ARB_shading_language_420pack GL_ARB_shading_language_include GL_ARB_shading_language_packing GL_ARB_spirv_extensions GL_ARB_stencil_texturing GL_ARB_sync GL_ARB_tessellation_shader GL_ARB_texture_barrier G
L_ARB_texture_buffer_object GL_ARB_texture_buffer_object_rgb32 GL_ARB_texture_buffer_range GL_ARB_texture_compression_bptc GL_ARB_texture_compression_rgtc GL_ARB_texture_cube_map_array GL_ARB_texture_filter_anisotropic GL_ARB_texture_float GL_ARB_texture_gather GL_ARB_texture_mirror_clamp_to_edge GL_ARB_texture_multisample GL_ARB_texture_non_power_of_two GL_ARB_texture_query_levels GL_ARB_texture_query_lod GL_ARB_texture_rectangle GL_ARB_texture_rg GL_ARB_texture_rgb10_a2ui GL_ARB_texture_stencil8 GL_ARB_texture_storage GL_ARB_texture_storage_multisample GL_ARB_texture_swizzle GL_ARB_texture_view GL_ARB_timer_query GL_ARB_transform_feedback2 GL_ARB_transform_feedback3 GL_ARB_transform_feedback_instanced GL_ARB_transform_feedback_overflow_query GL_ARB_uniform_buffer_object GL_ARB_vertex_array_bgra GL_ARB_vertex_array_object GL_ARB_vertex_attrib_64bit GL_ARB_vertex_attrib_binding GL_ARB_vertex_buffer_object GL_ARB_vertex_shader GL_ARB_vertex_type_10f_11f_11f_rev GL_ARB_vertex_type_2_10_10_10_rev GL_ARB_viewpor
t_array GL_ATI_blend_equation_separate GL_ATI_texture_float GL_EXT_EGL_image_storage GL_EXT_EGL_sync GL_EXT_abgr GL_EXT_blend_equation_separate GL_EXT_demote_to_helper_invocation GL_EXT_draw_buffers2 GL_EXT_draw_instanced GL_EXT_framebuffer_blit GL_EXT_framebuffer_multisample GL_EXT_framebuffer_multisample_blit_scaled GL_EXT_framebuffer_object GL_EXT_framebuffer_sRGB GL_EXT_packed_depth_stencil GL_EXT_packed_float GL_EXT_pixel_buffer_object GL_EXT_polygon_offset_clamp GL_EXT_provoking_vertex GL_EXT_semaphore GL_EXT_semaphore_fd GL_EXT_shader_framebuffer_fetch GL_EXT_shader_framebuffer_fetch_non_coherent GL_EXT_shader_integer_mix GL_EXT_shader_samples_identical GL_EXT_texture_array GL_EXT_texture_compression_dxt1 GL_EXT_texture_compression_rgtc GL_EXT_texture_compression_s3tc GL_EXT_texture_filter_anisotropic GL_EXT_texture_integer GL_EXT_texture_sRGB GL_EXT_texture_sRGB_R8 GL_EXT_texture_sRGB_decode GL_EXT_texture_shadow_lod GL_EXT_texture_shared_exponent GL_EXT_texture_snorm GL_EXT_texture_swizzle GL_EXT_ti
mer_query GL_EXT_transform_feedback GL_EXT_vertex_array_bgra GL_EXT_vertex_attrib_64bit GL_IBM_multimode_draw_arrays GL_INTEL_blackhole_render GL_INTEL_conservative_rasterization GL_INTEL_performance_query GL_INTEL_shader_atomic_float_minmax GL_INTEL_shader_integer_functions2 GL_KHR_blend_equation_advanced GL_KHR_blend_equation_advanced_coherent GL_KHR_context_flush_control GL_KHR_debug GL_KHR_no_error GL_KHR_parallel_shader_compile GL_KHR_robust_buffer_access_behavior GL_KHR_robustness GL_KHR_texture_compression_astc_ldr GL_KHR_texture_compression_astc_sliced_3d GL_MESA_framebuffer_flip_y GL_MESA_pack_invert GL_MESA_shader_integer_functions GL_MESA_texture_signed_rgba GL_NV_compute_shader_derivatives GL_NV_conditional_render GL_NV_copy_image GL_NV_depth_clamp GL_NV_fragment_shader_interlock GL_NV_packed_depth_stencil GL_NV_texture_barrier GL_OES_EGL_image GL_S3_s3tc
OPENGL LOG: Creating OpenGL 4.6 graphics device ; Context level  <OpenGL 4.5> ; Context handle 27192960
Begin MonoManager ReloadAssembly
- Completed reload, in  0.219 seconds
XRGeneral Settings awakening...

Default vsync count 0
requesting resize 128 x 128
resizing window to 128 x 128
Desktop is 1920 x 1080 @ 60 Hz
UnloadTime: 1.724553 ms
Setting up 6 worker threads for Enlighten.
  Thread -> id: 7f2230ba0700 -> priority: 1 
  Thread -> id: 7f21f3fff700 -> priority: 1 
  Thread -> id: 7f21f37fe700 -> priority: 1 
  Thread -> id: 7f21f2ffd700 -> priority: 1 
  Thread -> id: 7f21f27fc700 -> priority: 1 
  Thread -> id: 7f21f1ffb700 -> priority: 1 
requesting resize 256 x 256
resizing window to 256 x 256
Desktop is 1920 x 1080 @ 60 Hz
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/libKernel32
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/libKernel32.so
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/Kernel32
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/libKernel32
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/libKernel32.so
Fallback handler could not load library /home/johannes/tdw_build/TDW/TDW_Data/Mono/libKernel32
Waiting on threads to park on joinable thread list timed out.
```


Is it a problem that /is/sg/user/tdw_build/TDW/TDW_Data/Mono/ does not exist?

**Comment by meier-johannes94** [comment]
@alters-mit 

This is the first: 
```
(base) johannes@Johannes:~/Schreibtisch/tdw-test$ TRANSPORT_CHALLENGE=file:///~/Schreibtisch/tdw-test/model_library ./TDW/TDW.x86_64 -port 1071
Set current directory to /home/johannes/Schreibtisch/tdw-test
Found path: /home/johannes/Schreibtisch/tdw-test/TDW/TDW.x86_64
```


This is the second: 
```
(base) johannes@Johannes:~/Schreibtisch/tdw-test$ TRANSPORT_CHALLENGE=file:///~/Schreibtisch/tdw-test/model_library python test.py
Found path: /home/johannes/tdw_build/TDW/TDW.x86_64
/home/johannes/anaconda3/lib/python3.8/site-packages/sklearn/base.py:310: UserWarning: Trying to unpickle estimator GaussianMixture from version 0.22.2.post1 when using version 0.24.1. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
Controller connected
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    obs, info = env.reset(dataset[scene_number])
  File "/home/johannes/Schreibtisch/tdw-test/pyastar/tdw-transport-challenge/tdw_transport_challenge/tdw_gym.py", line 116, in reset
    self.controller.init_scene(scene=scene, layout=layout, room = 5, random_seed=seed)
  File "/home/johannes/.local/share/Trash/files/transport_challenge/transport_challenge/transport_challenge/transport_challenge/transport_controller.py", line 223, in init_scene
    return super().init_scene(scene=scene, layout=layout, room=room)
  File "/home/johannes/anaconda3/lib/python3.8/site-packages/magnebot/magnebot_controller.py", line 424, in init_scene
    self._cache_static_data(resp=resp)
  File "/home/johannes/.local/share/Trash/files/transport_challenge/transport_challenge/transport_challenge/transport_challenge/transport_controller.py", line 653, in _cache_static_data
    super()._cache_static_data(resp=resp)
  File "/home/johannes/anaconda3/lib/python3.8/site-packages/magnebot/magnebot_controller.py", line 1855, in _cache_static_data
    for i in range(segmentation_colors.get_num()):
AttributeError: 'NoneType' object has no attribute 'get_num'

```
Both are on the private (=second) machine. Two windows open up additionally. But they show basically black content. I once managed hat it once shows some kind of scene, but the scene was static. When I run those commands, then it is very computationally very demaning for the PC.

**Comment by instbldrjems** [comment]
@meier-johannes94 No, this is not normal. When we say you need to wait for the scene to load, we mean a few minutes (maybe 5 minutes) the first time you run it.  

Are you sure you did not have another instance of the TDW build running, that you did not kill, before running this?  The errors at the bottom of the window on the left of this image are typical of trying to connect to a build when there is already an existing connection.  

@alters-mit will take a look at this first thing tomorrow. However If it would help, we could set up a zoom call with you tomorrow to try and debug this together? Let me know via email if you would like to do that -- jeremyes@nnit.edu

**Comment by abhi1092** [comment]
@meier-johannes94 I have pushed a few changes to the transport challenge [gym code](https://github.com/chuangg/tdw-transport-challenge-starter-code#:~:text=Transport%20Challenge%20Gym%20API). Please pull the new changes. The 3d model library we provided was missing models for osx and windows platforms. I have updated that too, just re-download the data. Once the model library is downloaded and `TRANSPORT_CHALLENGE` environment variable is set the environment 
 loading should not take a lot of time. 
And there were some minor fixes to `test.py` script. 
The testing script uses `h_agent` to navigate the scene. This is just a baseline agent we provided as an example and is not required for using the gym or for the building and submitting code for transport challenge.  The `h_agent` code has some bugs and might error out randomly. But the base gym code is working.

**Comment by meier-johannes94** [comment]
@abhi1092 
Thank you! I got the new files and indeed it seems faster. It would be great if h_agent becomes functional again. :-)
I have a few theoretical questions about your approach (in the paper). I think maybe it is best I open a new issue. I look forward to your answer! 

For now I think this issue is solved.
